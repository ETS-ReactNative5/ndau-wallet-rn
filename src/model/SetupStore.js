import { Alert } from 'react-native';
import AppConstants from '../AppConstants';

class SetupStore {
  constructor() {
    if (!SetupStore.instance) {
      SetupStore.instance = this;
    }

    this._userId = '';
    this._numberOfAccounts = 0;
    this._qrCode = '';
    this._encryptionPassword = '';
    this._entropy = '';
    this._recoveryPhrase = '';
    this._shuffledMap = [];
    this._shuffledWord = [];
    this._addressType = AppConstants.MAINNET_ADDRESS;
    this._walletName = '';
  }

  set userId(userId) {
    this._userId = userId;
    this._printData();
  }

  get userId() {
    console.debug(`SetupStore.userId ${this._userId}`);
    return this._userId;
  }

  set numberOfAccounts(numberOfAccounts) {
    this._numberOfAccounts = numberOfAccounts;
    this._printData();
  }

  get numberOfAccounts() {
    console.debug(`SetupStore.numberOfAccounts ${this._numberOfAccounts}`);
    return this._numberOfAccounts;
  }

  set qrCode(qrCode) {
    this._qrCode = qrCode;
    this._printData();
  }

  get qrCode() {
    console.debug(`SetupStore.qrCode ${this._qrCode}`);
    return this._qrCode;
  }

  set encryptionPassword(encryptionPassword) {
    this._encryptionPassword = encryptionPassword;
    this._printData();
  }

  get encryptionPassword() {
    console.debug(`SetupStore.encryptionPassword ${this._encryptionPassword}`);
    return this._encryptionPassword;
  }

  set entropy(entropy) {
    this._entropy = entropy;
    this._printData();
  }

  get entropy() {
    console.debug(`SetupStore.entropy ${this._entropy}`);
    return this._entropy;
  }

  set recoveryPhrase(recoveryPhrase) {
    this._recoveryPhrase = recoveryPhrase.slice();
    this._printData();
  }

  get recoveryPhrase() {
    console.debug(`SetupStore.recoveryPhrase ${this._recoveryPhrase}`);
    return this._recoveryPhrase;
  }

  set shuffledWords(shuffledWords) {
    this._shuffledWords = shuffledWords.slice();
    this._printData();
  }

  get shuffledWords() {
    console.debug(`SetupStore.shuffledWords ${this._shuffledWords}`);
    return this._shuffledWords;
  }

  set shuffledMap(shuffledMap) {
    this._shuffledMap = shuffledMap.slice();
    this._printData();
  }

  get shuffledMap() {
    console.debug(`SetupStore.shuffledMap ${this._shuffledMap}`);
    return this._shuffledMap;
  }

  toggleAddressType() {
    const oldAddressType = this._addressType;
    const newAddressType =
      this._addressType === AppConstants.MAINNET_ADDRESS
        ? AppConstants.TESTNET_ADDRESS
        : AppConstants.MAINNET_ADDRESS;

    Alert.alert(
      'Information',
      `Old address type was ${oldAddressType} which has been moved to ${newAddressType}`,
      [{ text: 'OK', onPress: () => { } }],
      { cancelable: false }
    );

    this._addressType = newAddressType;
    this._printData();
  }

  get addressType() {
    console.debug(`SetupStore.addressType ${this._addressType}`);
    return this._addressType;
  }

  set walletName(walletName) {
    this._walletName = walletName;
    this._printData();
  }

  get walletName() {
    console.debug(`SetupStore.walletName ${this._walletName}`);
    return this._walletName;
  }

  _printData = () => {
    // console.debug(`SetupStore.userId ${this._userId}`);
    // console.debug(`SetupStore.numberOfAccounts ${this._numberOfAccounts}`);
    // console.debug(`SetupStore.qrCode ${this._qrCode}`);
    // console.debug(`SetupStore.entropy ${this._entropy}`);
    // console.debug(`SetupStore.recoveryPhrase ${this._recoveryPhrase}`);
    // console.debug(`SetupStore.shuffledWords ${this._shuffledWords}`);
    // console.debug(`SetupStore.shuffledMap ${this._shuffledMap}`);
    // console.debug(`SetupStore.walletName ${this._walletName}`);
    // console.debug(`SetupStore.addressType ${this._addressType}`);
  };
}

const instance = new SetupStore();

export default instance;
