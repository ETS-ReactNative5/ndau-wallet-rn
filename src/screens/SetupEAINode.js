import React, { Component } from 'react';
import { StyleSheet, View, ScrollView, Text, SafeAreaView, NativeModules } from 'react-native';
import ndauApi from '../api/NdauAPI';
import AsyncStorageHelper from '../model/AsyncStorageHelper';
import CommonButton from '../components/CommonButton';
import Stepper from '../components/Stepper';
import { Dropdown } from 'react-native-material-dropdown';
import cssStyles from '../css/styles';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { push, setUser, startTabBasedApp } from '../actions/NavigationActions';

//"nd" for mainnet, or "tn" for testnet.
const addressGenerationType = 'nd';

class SetupEAINode extends Component {
  constructor(props) {
    super(props);
    const nodeNames = [
      {
        value: 'Boylston'
      },
      {
        value: 'Newbury'
      },
      {
        value: 'Commonwealth'
      },
      {
        value: 'Dartmouth'
      },
      {
        value: 'Storrow'
      }
    ];

    this.state = {
      node: '',
      data: nodeNames,
      selectedNode: nodeNames[Math.floor(Math.random() * nodeNames.length)].value
    };
  }

  sendAccountAddresses = (userId, addresses, token) => {
    return new Promise((resolve, reject) => {
      ndauApi
        .sendAccountAddresses(userId, addresses, token)
        .then((whatPersisted) => {
          console.debug(`sendAccountAddresses persisted: ${whatPersisted}`);
          resolve(whatPersisted);
        })
        .catch((error) => {
          console.error(error);
          reject(error);
        });
    });
  };

  finishSetup = async () => {
    console.debug('Finishing Setup...');
    const addresses = await this.keyGeneration();

    this.sendAddressesToOneiro(addresses)
      .then(() => {
        this.persistAddresses(addresses);

        this.props.startTabBasedApp();
      })
      .catch((error) => {
        console.error(error);
      });
  };

  keyGeneration = async () => {
    console.debug('Generating all keys from phrase given...');
    const seedPhraseString = this.props.reduxProps.seedPhrase.join().replace(/,/g, ' ');
    console.debug(`seedPhraseString: ${seedPhraseString}`);
    const seedPhraseAsBytes = await NativeModules.KeyaddrManager.KeyaddrWordsToBytes(
      'en',
      seedPhraseString
    );
    console.debug(`seedPhraseAsBytes: ${seedPhraseAsBytes}`);
    const publicAddresses = await NativeModules.KeyaddrManager.CreatePublicAddress(
      seedPhraseAsBytes,
      this.props.reduxProps.numberOfAccounts,
      addressGenerationType
    );
    console.debug(`publicAddresses: ${publicAddresses}`);

    return publicAddresses;
  };

  sendAddressesToOneiro = (addresses) => {
    return this.sendAccountAddresses(
      this.props.reduxProps.userId,
      addresses,
      this.props.reduxProps.qrCode
    );
  };

  persistAddresses = (addresses) => {
    const user = {
      userId: this.props.reduxProps.userId,
      addresses: addresses,
      selectedNode: this.state.selectedNode
    };
    AsyncStorageHelper.setUser(user, this.props.reduxProps.encryptionPassword);
    this.props.setUser(user);
  };

  render() {
    return (
      <SafeAreaView style={styles.safeContainer}>
        <View style={cssStyles.container}>
          <ScrollView style={styles.contentContainer}>
            <Stepper screenNumber={8} />
            <View style={styles.textContainer}>
              <Text style={cssStyles.wizardText}>
                In order to earn your Ecosystem Alignment Incentive (EAI) you must delegate your
                ndau to a node. Please select the default node for your accounts. You will be able
                to change this later.
              </Text>
            </View>
            <Dropdown
              label="Please choose an EAI node"
              data={this.state.data}
              baseColor="#ffffff"
              selectedItemColor="#000000"
              textColor="#ffffff"
              itemTextStyle={cssStyles.wizardText}
              fontSize={20}
              labelFontSize={14}
              value={this.state.selectedNode}
              onChangeText={(selectedNode) => this.setState({ selectedNode })}
            />
          </ScrollView>
          <View style={styles.footer}>
            <CommonButton
              onPress={this.finishSetup}
              title="Select and Finish"
              id="select-and-finish"
            />
          </View>
        </View>
      </SafeAreaView>
    );
  }
}

const styles = StyleSheet.create({
  safeContainer: {
    flex: 1,
    backgroundColor: '#1c2227'
  },
  button: {
    marginTop: 0
  },
  textContainer: {
    marginBottom: 5
  },
  contentContainer: {
    flex: 1 // pushes the footer to the end of the screen
  },
  footer: {
    justifyContent: 'flex-end'
  },
  progress: {
    paddingTop: 30,
    paddingBottom: 30
  },
  checkbox: { flex: 1, padding: 10 }
});

const mapStateToProps = (state) => {
  return {
    userId: state.userId,
    qrCode: state.qrCode,
    numberOfAccounts: state.numberOfAccounts,
    seedPhrase: state.seedPhrase
  };
};

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({ push, setUser, startTabBasedApp }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(SetupEAINode);
