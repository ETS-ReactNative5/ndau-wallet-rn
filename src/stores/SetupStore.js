import { Alert } from 'react-native'
import AppConstants from '../AppConstants'
import LoggingService from '../services/LoggingService'

class SetupStore {
  constructor () {
    if (!SetupStore.instance) {
      SetupStore.instance = this
    }

    this._userId = ''
    this._numberOfAccounts = 0
    this._qrCode = ''
    this._encryptionPassword = ''
    this._entropy = ''
    this._recoveryPhrase = ''
    this._shuffledMap = []
    this._shuffledWord = []
    this._addressType = AppConstants.MAINNET_ADDRESS
    this._walletId = ''
  }

  set userId (userId) {
    this._userId = userId
  }

  get userId () {
    LoggingService.debug(`SetupStore.userId ${this._userId}`)
    return this._userId
  }

  set numberOfAccounts (numberOfAccounts) {
    this._numberOfAccounts = numberOfAccounts
  }

  get numberOfAccounts () {
    LoggingService.debug(`SetupStore.numberOfAccounts ${this._numberOfAccounts}`)
    return this._numberOfAccounts
  }

  set qrCode (qrCode) {
    this._qrCode = qrCode
  }

  get qrCode () {
    LoggingService.debug(`SetupStore.qrCode ${this._qrCode}`)
    return this._qrCode
  }

  set encryptionPassword (encryptionPassword) {
    this._encryptionPassword = encryptionPassword
  }

  get encryptionPassword () {
    LoggingService.debug(`SetupStore.encryptionPassword ${this._encryptionPassword}`)
    return this._encryptionPassword
  }

  set entropy (entropy) {
    this._entropy = entropy
  }

  get entropy () {
    LoggingService.debug(`SetupStore.entropy ${this._entropy}`)
    return this._entropy
  }

  set recoveryPhrase (recoveryPhrase) {
    this._recoveryPhrase = recoveryPhrase.slice()
  }

  get recoveryPhrase () {
    LoggingService.debug(`SetupStore.recoveryPhrase ${this._recoveryPhrase}`)
    return this._recoveryPhrase
  }

  set shuffledWords (shuffledWords) {
    this._shuffledWords = shuffledWords.slice()
  }

  get shuffledWords () {
    LoggingService.debug(`SetupStore.shuffledWords ${this._shuffledWords}`)
    return this._shuffledWords
  }

  set shuffledMap (shuffledMap) {
    this._shuffledMap = shuffledMap.slice()
  }

  get shuffledMap () {
    LoggingService.debug(`SetupStore.shuffledMap ${this._shuffledMap}`)
    return this._shuffledMap
  }

  toggleAddressType () {
    const oldAddressType = this._addressType
    const newAddressType =
      this._addressType === AppConstants.MAINNET_ADDRESS
        ? AppConstants.TESTNET_ADDRESS
        : AppConstants.MAINNET_ADDRESS

    Alert.alert(
      'Information',
      `Old address type was ${oldAddressType} which has been moved to ${newAddressType}`,
      [{ text: 'OK', onPress: () => {} }],
      { cancelable: false }
    )

    this._addressType = newAddressType
  }

  get addressType () {
    LoggingService.debug(`SetupStore.addressType ${this._addressType}`)
    return this._addressType
  }

  set walletId (walletId) {
    this._walletId = walletId
  }

  get walletId () {
    LoggingService.debug(`SetupStore.walletId ${this._walletId}`)
    return this._walletId
  }

  printData = () => {
    LoggingService.debug(`SetupStore.userId ${this._userId}`)
    LoggingService.debug(`SetupStore.numberOfAccounts ${this._numberOfAccounts}`)
    LoggingService.debug(`SetupStore.qrCode ${this._qrCode}`)
    LoggingService.debug(`SetupStore.entropy ${this._entropy}`)
    LoggingService.debug(`SetupStore.recoveryPhrase ${this._recoveryPhrase}`)
    LoggingService.debug(`SetupStore.shuffledWords ${this._shuffledWords}`)
    LoggingService.debug(`SetupStore.shuffledMap ${this._shuffledMap}`)
    LoggingService.debug(`SetupStore.walletId ${this._walletId}`)
    LoggingService.debug(`SetupStore.addressType ${this._addressType}`)
  }

  reset = () => {
    this._userId = ''
    this._numberOfAccounts = 0
    this._qrCode = ''
    this._encryptionPassword = ''
    this._entropy = ''
    this._recoveryPhrase = ''
    this._shuffledMap = []
    this._shuffledWord = []
    this._addressType = AppConstants.MAINNET_ADDRESS
    this._walletId = ''
  }
}

const instance = new SetupStore()

export default instance
