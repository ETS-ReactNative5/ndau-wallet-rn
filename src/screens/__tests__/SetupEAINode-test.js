import { mount } from 'enzyme';
import React from 'react';
import { NativeModules } from 'react-native';
import SetupEAINode from '../SetupEAINode';
import sinon from 'sinon';
import renderer from 'react-test-renderer';
import SetupStore from '../../model/SetupStore';

jest.mock('NativeModules', () => {
  return {
    KeyaddrManager: {
      KeyaddrWordsToBytes: jest.fn(),
      CreatePublicAddress: jest.fn()
    },
    KeyboardObserver: {}
  };
});

describe('testing SetupEAINode...', () => {
  let seedPhraseArray = [
    'goat',
    'amount',
    'liar',
    'amount',
    'expire',
    'adjust',
    'cage',
    'candy',
    'arch',
    'gather',
    'drum',
    'buyer'
  ];
  let userId = 'TAC-3PY';
  let numberOfAccounts = 5;
  const bytes = 'ZWEQAwQFBgcICQoLDA0ODw==';
  const KeyaddrWordsToBytes = sinon.spy(NativeModules.KeyaddrManager, 'KeyaddrWordsToBytes');
  KeyaddrWordsToBytes.mockReturnValue(bytes);
  const CreatePublicAddress = sinon.spy(NativeModules.KeyaddrManager, 'CreatePublicAddress');
  CreatePublicAddress.mockReturnValue([]);
  const navigation = {
    navigate: () => {}
  };
  SetupStore.setSeedPhrase(seedPhraseArray);
  SetupStore.setUserId(userId);
  SetupStore.setNumberOfAccounts(numberOfAccounts);
  SetupStore.setQRCode(bytes);

  beforeAll(() => {
    this.oldRandom = global.Math.random;
    const mockMathRandom = Object.create(global.Math);
    mockMathRandom.random = () => 0.5;
    global.Math = mockMathRandom;
  });

  afterAll(() => {
    global.Math.random = this.oldRandom;
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <SetupEAINode
          seedPhraseArray={seedPhraseArray}
          userId={userId}
          numberOfAccounts={numberOfAccounts}
          navigation
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  //THIS NEED TO BE FIXED BUT IT CAN'T AT PRESENT
  // it('finishes setup successfully', async () => {
  //   const wrapper = mount(
  //     <SetupEAINode
  //       seedPhraseArray={seedPhraseArray}
  //       userId={userId}
  //       numberOfAccounts={numberOfAccounts}
  //       navigation
  //     />
  //   );

  //   expect(KeyaddrWordsToBytes.calledOnce).toBe(false);
  //   expect(CreatePublicAddress.calledOnce).toBe(false);

  //   const onlyButton = wrapper.find('#select-and-finish');
  //   console.log(`onlyButton is: ${JSON.stringify(onlyButton)}`);
  //   expect(onlyButton.length).toBe(1);
  //   await onlyButton.simulate('press');

  //   console.log(`stuff: ${JSON.stringify(KeyaddrWordsToBytes.mock, null, 2)}`);

  //   expect(KeyaddrWordsToBytes.mock.results[0].value).toBe(bytes);
  //   expect(CreatePublicAddress.mock.results[0].value).toEqual([]);
  // });
});
