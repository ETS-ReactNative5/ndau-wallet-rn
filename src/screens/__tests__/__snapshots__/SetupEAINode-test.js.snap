// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing SetupEAINode... renders correctly 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "#1c2227",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#1c2227",
        "flex": 1,
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontSize": 20,
              }
            }
          >
            Select a node
          </Text>
        </View>
        <View>
          <RCTProgressView
            progress={1}
            style={
              Array [
                Object {
                  "height": 2,
                },
                undefined,
              ]
            }
          />
        </View>
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontSize": 20,
              }
            }
          >
            In order to earn your Ecosystem Alignment Incentive (EAI) you must delegate your ndau to a node. Please select the default node for your accounts. You will be able to change this later.
          </Text>
        </View>
        <View>
          <RCTPicker
            items={
              Array [
                Object {
                  "label": "Node A",
                  "textColor": undefined,
                  "value": "nodeA",
                },
                Object {
                  "label": "Node B",
                  "textColor": undefined,
                  "value": "nodeB",
                },
              ]
            }
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            selectedIndex={0}
            style={
              Array [
                Object {
                  "height": 216,
                },
                Object {
                  "borderColor": "gray",
                  "borderWidth": 1,
                  "color": "#ffffff",
                  "height": 50,
                  "marginBottom": 10,
                  "marginTop": 10,
                  "paddingLeft": 10,
                },
              ]
            }
          />
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#4d9678",
                },
              ]
            }
          >
            Select and finish
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
